File: experiments/neural-showcase-v3/src/lib/meeet.ts
/**
 * Client + React hook for the meeet durable bridge.
 *
 * Surfaces:
 * - /api/meeet/health → bridge state (ingest URL, store stats,
 *   last replay metadata).
 * - /api/meeet/events → newest-first event list with filters.
 * - /api/meeet/replay → manual flush trigger.
 *
 * The cockpit renders this as a "black box" panel with a small
 * status indicator and an event timeline.
 */

import { useCallback, useEffect, useRef, useState } from "react";

import { API_BASE } from "./api";

export interface MeeetEvent {
  id: number;
  ts: number;
  trace_id: string;
  /** Optional — present once K-class events started carrying it. */
  session_id?: string | null;
  /** ``edge`` / ``cloud`` / ``fallback`` / ``mixed`` (or null). */
  route?: string | null;
  kind: string;
  source: string;
  contract_version: string;
  payload: Record<string, unknown>;
  pushed: boolean;
  pushed_at: number | null;
  last_error: string | null;
}

/** Trace-summary row matches ``backend/core/meeet/trace_summary.py::TraceSummary.to_dict()``. */
export interface TraceSummary {
  trace_id: string;
  event_count: number;
  kinds: string[];
  routes: string[];
  primary_route: string | null;
  total_cost_usd: number;
  tokens_in: number;
  tokens_out: number;
  contradictions: number;
  error_count: number;
  last_session_id: string | null;
  started_at: number | null;
  ended_at: number | null;
  duration_ms: number | null;
  updated_at: number;
}

export interface MeeetStats {
  total: number;
  unpushed: number;
  first_ts: number | null;
  last_ts: number | null;
  db_path: string | null;
  enabled: boolean;
}

export interface LastReplay {
  enabled: boolean;
  pushed?: number;
  failed?: number;
  scanned?: number;
  remaining?: number;
  ran_at?: number;
}

export interface MeeetHealth {
  ok: boolean;
  client: {
    enabled: boolean;
    ingest_url: string | null;
    api_key_set: boolean;
    contract_version: string;
    source: string;
  };
  store: MeeetStats;
  last_replay: LastReplay | null;
}

export async function getMeeetHealth(): Promise<MeeetHealth> {
  const r = await fetch(`${API_BASE}/api/meeet/health`);
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return r.json() as Promise<MeeetHealth>;
}

export async function listEvents(
  opts: {
    limit?: number;
    since?: number;
    trace_id?: string;
    kind?: string;
    only_unpushed?: boolean;
  } = {},
): Promise<MeeetEvent[]> {
  const params = new URLSearchParams();
  if (opts.limit) params.set("limit", String(opts.limit));
  if (opts.since) params.set("since", String(opts.since));
  if (opts.trace_id) params.set("trace_id", opts.trace_id);
  if (opts.kind) params.set("kind", opts.kind);
  if (opts.only_unpushed) params.set("only_unpushed", "true");
  const qs = params.toString();
  const r = await fetch(
    `${API_BASE}/api/meeet/events${qs ? `?${qs}` : ""}`,
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const d = (await r.json()) as { events: MeeetEvent[] };
  return d.events ?? [];
}

export async function replayNow(
  limit = 100,
): Promise<LastReplay> {
  const r = await fetch(
    `${API_BASE}/api/meeet/replay?limit=${limit}`,
    { method: "POST" },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return r.json() as Promise<LastReplay>;
}

/**
 * GET /api/meeet/traces — newest-first rolled-up trace summaries.
 *
 * Server defaults: limit 50 (max 500). Filters: since (epoch s),
 * primary_route, session_id.
 */
export async function listTraces(
  opts: {
    limit?: number;
    since?: number;
    primary_route?: string;
    session_id?: string;
  } = {},
): Promise<TraceSummary[]> {
  const params = new URLSearchParams();
  if (opts.limit) params.set("limit", String(opts.limit));
  if (opts.since != null) params.set("since", String(opts.since));
  if (opts.primary_route) params.set("primary_route", opts.primary_route);
  if (opts.session_id) params.set("session_id", opts.session_id);
  const qs = params.toString();
  const r = await fetch(
    `${API_BASE}/api/meeet/traces${qs ? `?${qs}` : ""}`,
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const d = (await r.json()) as { traces: TraceSummary[] };
  return d.traces ?? [];
}

/** GET /api/meeet/traces/{trace_id} — single rollup row, 404s on miss. */
export async function getTrace(traceId: string): Promise<TraceSummary | null> {
  const r = await fetch(
    `${API_BASE}/api/meeet/traces/${encodeURIComponent(traceId)}`,
  );
  if (r.status === 404) return null;
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const d = (await r.json()) as { trace?: TraceSummary };
  return d.trace ?? null;
}

/** POST /api/meeet/traces/refresh — operator-initiated rebuild. */
export async function refreshTraces(
  opts: { since?: number } = {},
): Promise<{ ok: boolean; rebuilt: number }> {
  const params = new URLSearchParams();
  if (opts.since != null) params.set("since", String(opts.since));
  const qs = params.toString();
  const r = await fetch(
    `${API_BASE}/api/meeet/traces/refresh${qs ? `?${qs}` : ""}`,
    { method: "POST" },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return r.json();
}

/**
 * Polling hook for the trace summaries list. Fits the cockpit's
 * "ambient telemetry" pattern: a five-second refresh that doesn't
 * thrash the React tree (single state object, ref-guarded against
 * unmount races).
 */
export function useTraceSummaries(
  opts: {
    limit?: number;
    intervalMs?: number;
    since?: number;
    primary_route?: string;
    session_id?: string;
  } = {},
): {
  traces: TraceSummary[];
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
} {
  const [traces, setTraces] = useState<TraceSummary[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const cancelled = useRef(false);
  const intervalMs = opts.intervalMs ?? 5000;

  const refresh = useCallback(async () => {
    try {
      const list = await listTraces({
        limit: opts.limit ?? 100,
        since: opts.since,
        primary_route: opts.primary_route,
        session_id: opts.session_id,
      });
      if (!cancelled.current) {
        setTraces(list);
        setError(null);
      }
    } catch (e) {
      if (!cancelled.current) setError((e as Error).message);
    } finally {
      if (!cancelled.current) setLoading(false);
    }
  }, [opts.limit, opts.since, opts.primary_route, opts.session_id]);

  useEffect(() => {
    cancelled.current = false;
    void refresh();
    if (intervalMs > 0) {
      const id = window.setInterval(refresh, intervalMs);
      return () => {
        cancelled.current = true;
        window.clearInterval(id);
      };
    }
    return () => {
      cancelled.current = true;
    };
  }, [refresh, intervalMs]);

  return { traces, loading, error, refresh };
}

export function useMeeetHealth(intervalMs = 5000): {
  health: MeeetHealth | null;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
} {
  const [health, setHealth] = useState<MeeetHealth | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const cancelled = useRef(false);

  const refresh = useCallback(async () => {
    try {
      const out = await getMeeetHealth();
      if (!cancelled.current) {
        setHealth(out);
        setError(null);
      }
    } catch (e) {
      if (!cancelled.current) setError((e as Error).message);
    } finally {
      if (!cancelled.current) setLoading(false);
    }
  }, []);

  useEffect(() => {
    cancelled.current = false;
    void refresh();
    if (intervalMs > 0) {
      const id = window.setInterval(refresh, intervalMs);
      return () => {
        cancelled.current = true;
        window.clearInterval(id);
      };
    }
    return () => {
      cancelled.current = true;
    };
  }, [refresh, intervalMs]);

  return { health, loading, error, refresh };
}

/**
 * Rolling event timeline — pulls every `intervalMs` (default 5s).
 * Pass `kind` or `traceId` to filter on the server side.
 */
export function useMeeetEvents(
  opts: {
    limit?: number;
    intervalMs?: number;
    kind?: string;
    traceId?: string;
    onlyUnpushed?: boolean;
  } = {},
): {
  events: MeeetEvent[];
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
} {
  const [events, setEvents] = useState<MeeetEvent[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const cancelled = useRef(false);
  const intervalMs = opts.intervalMs ?? 5000;

  const refresh = useCallback(async () => {
    try {
      const list = await listEvents({
        limit: opts.limit ?? 200,
        kind: opts.kind,
        trace_id: opts.traceId,
        only_unpushed: opts.onlyUnpushed,
      });
      if (!cancelled.current) {
        setEvents(list);
        setError(null);
      }
    } catch (e) {
      if (!cancelled.current) setError((e as Error).message);
    } finally {
      if (!cancelled.current) setLoading(false);
    }
  }, [opts.limit, opts.kind, opts.traceId, opts.onlyUnpushed]);

  useEffect(() => {
    cancelled.current = false;
    void refresh();
    if (intervalMs > 0) {
      const id = window.setInterval(refresh, intervalMs);
      return () => {
        cancelled.current = true;
        window.clearInterval(id);
      };
    }
    return () => {
      cancelled.current = true;
    };
  }, [refresh, intervalMs]);

  return { events, loading, error, refresh };
}

File: web_extras/routers/meeet.py
"""HTTP surface over the meeet durable store.

Endpoints:

- ``GET /api/meeet/stats`` — quick summary (total / unpushed / first_ts / last_ts).
- ``GET /api/meeet/events`` — list events newest-first with filters
  ``limit, since, trace_id, kind, kind_prefix, session_id, only_unpushed``.
- ``POST /api/meeet/replay`` — flush unpushed events to ingest now.
- ``GET /api/meeet/health`` — bridge health (ingest url set?, api key set?,
  contract version, store stats, last replay attempt).
- ``GET /api/meeet/traces`` — list rolled-up trace summaries
  (event_count, kinds, route, cost, contradictions, duration_ms).
- ``GET /api/meeet/traces/{trace_id}`` — single rollup row.
- ``POST /api/meeet/traces/refresh`` — rebuild the rollup from
  the events table on demand.

Every event TARS emits flows through the SQLite WAL store, so this is
the operator's local "black box": survives offline, replays on
reconnect.
"""

from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query

from backend.core.meeet import (
    get_client,
    get_store,
    get_trace_summary_store,
)

router = APIRouter(prefix="/api/meeet", tags=["meeet"])


@router.get("/stats")
async def stats() -> dict[str, Any]:
    return await get_store().stats()


@router.get("/events")
async def list_events(
    limit: int = Query(default=100, ge=1, le=1000),
    since: Optional[float] = Query(default=None),
    trace_id: Optional[str] = Query(default=None),
    kind: Optional[str] = Query(default=None),
    kind_prefix: Optional[str] = Query(
        default=None,
        description=(
            "Filter to events whose kind starts with this prefix "
            "(e.g. 'pair.' for the pairing audit lane)."
        ),
        max_length=64,
    ),
    session_id: Optional[str] = Query(default=None),
    only_unpushed: bool = Query(default=False),
) -> dict[str, Any]:
    events = await get_store().list_events(
        limit=limit,
        since=since,
        trace_id=trace_id,
        kind=kind,
        kind_prefix=kind_prefix,
        session_id=session_id,
        only_unpushed=only_unpushed,
    )
    return {
        "ok": True,
        "count": len(events),
        "events": [
            {
                "id": e.id,
                "ts": e.ts,
                "trace_id": e.trace_id,
                "session_id": e.session_id,
                "route": e.route,
                "kind": e.kind,
                "source": e.source,
                "contract_version": e.contract_version,
                "payload": e.payload,
                "pushed": e.pushed,
                "pushed_at": e.pushed_at,
                "last_error": e.last_error,
            }
            for e in events
        ],
    }


@router.post("/replay")
async def replay(limit: int = Query(default=100, ge=1, le=1000)) -> dict[str, Any]:
    return await get_client().replay_unpushed(limit=limit)


@router.get("/health")
async def health() -> dict[str, Any]:
    return await get_client().health()


@router.get("/traces")
async def list_traces(
    limit: int = Query(default=50, ge=1, le=500),
    since: Optional[float] = Query(default=None),
    primary_route: Optional[str] = Query(default=None),
    session_id: Optional[str] = Query(default=None),
) -> dict[str, Any]:
    summaries = await get_trace_summary_store().list_summaries(
        limit=limit,
        since=since,
        primary_route=primary_route,
        session_id=session_id,
    )
    return {
        "ok": True,
        "count": len(summaries),
        "traces": [s.to_dict() for s in summaries],
    }


@router.get("/traces/{trace_id}")
async def get_trace(trace_id: str) -> dict[str, Any]:
    summary = await get_trace_summary_store().get(trace_id)
    if summary is None:
        raise HTTPException(status_code=404, detail="trace_not_found")
    return {"ok": True, "trace": summary.to_dict()}


@router.post("/traces/refresh")
async def refresh_traces(
    since: Optional[float] = Query(default=None),
) -> dict[str, Any]:
    return await get_trace_summary_store().rebuild(since=since)
